### ISSUE TITLE
Disallow string input to API functions

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#11003]
Most SymPy functions "accidentally" accept strings as input, because they call `sympify` on the input, and sympify converts strings to SymPy expressions. 

This would be somewhat of an API break, but I think we should disallow this. The problems with strings as input are outlined [here](https://github.com/sympy/sympy/wiki/Idioms-and-Antipatterns#strings-as-input). Additionally, it only works by accident, and for some functions, it doesn't work at all. 

I think we should remove it. We can use `_sympify` to convert expressions, which doesn't allow strings. Perhaps there's a better way, which prevents `sympify` from creeping back in. 

This is technically an API break, so we should consider if we want to run a deprecation cycle. It's not documented API, and I at least personally have stated several times that this could change at any time. Regardless, we should make the error message for string input clear that strings can be converted to SymPy expressions via `sympify`. We may also want to copy the contents of [that wiki page](https://github.com/sympy/sympy/wiki/Idioms-and-Antipatterns#strings-as-input) to the official docs.

**Comment by asmeurer** [comment]
This will be easy to fix once we decide how we want to do it.

**Comment by oscarbenjamin** [comment]
I agree. String input should be disallowed. I see a lot of examples using this though so maybe it does need a deprecation cycle. 

Actually I'd rather not allow string input to simpify either. There should be a separate function for parsing strings as expressions since it is an entirely different operation.

We should consider and document the alternatives carefully either way.

**Comment by oscarbenjamin** [comment]
Since we're now talking about milestones for 1.4 I think that some kind of deprecation for this should go in there.

**Comment by asmeurer** [comment]
> I agree. String input should be disallowed. I see a lot of examples using this though so maybe it does need a deprecation cycle.

I've always considered functions supporting strings to be accidental behavior (with a couple of exceptions) that shouldn't be relied upon. I wouldn't be opposed to removing it without a deprecation, but maybe we should do it. That's what I wrote [here](https://github.com/sympy/sympy/wiki/Idioms-and-Antipatterns#strings-as-input) (though that isn't technically part of the documentation). Then again my experience from things like StackOverflow and looking around GitHub for uses of SymPy is that a *lot* of users use string inputs, so maybe a deprecation is better. 

By the way, another reason to use `_sympify` instead of `sympify` is that it can have huge performance implications (https://github.com/sympy/sympy/pull/15769). That is slightly different because it's about `sympify` converting non-string objects into strings and parsing them, but it's effectively the same because both that and direct parsing of strings are controlled by the same flag to `sympify` (`strict=True`, or `_sympify` which is the same thing).

> Actually I'd rather not allow string input to simpify either. There should be a separate function for parsing strings as expressions since it is an entirely different operation.

There is `parse_expr`. But `sympify` has parsed strings since forever. It's one of its main uses. So I'm not sure I agree that `strict=True` should be made the default. It would be a good idea to add `parse_expr` to the main `__init__.py`, however. 

> Since we're now talking about milestones for 1.4 I think that some kind of deprecation for this should go in there.

Maybe if you can pull something together by then. I was hoping to get the release branch started this week. I've already postponed it for a while (I originally wanted to get it out last year). 

I think changing the default for `sympify` itself should be discussed more. That means that to implement this for now at least, every function in SymPy that calls `sympify` on its input should use `_sympify` instead, or a deprecation wrapper if we decide to go that route. It might also be good to somehow test this. There are also some SymPy functions that don't call sympify at all, but they probably should be (if they would fail on non-sympified `int` inputs they should be calling it).

**Comment by oscarbenjamin** [comment]
Okay maybe this needs more discussion if the release is branching in a week.

Might not happen for 1.4 but I propose the following:

1. Add a sympify_deprecated function that warns for string inputs and then calls sympify (or parse_expr/_sympify).
2. All public APIs that use sympify start using sympify_deprecated in 1.4.
3. The warning should recommend users explicitly use sympify or parse_expr.
4. In say 1.5 or 1.6 sympify_deprecated is replaced with _sympify.

**Comment by asmeurer** [comment]
I think we should do a deprecation. There is so much user code that does this. We might even want to deprecate for a while, so that it can get ingrained in the general consciousness of SymPy users that string inputs are bad. Actually, the warning should recommend for most users to create expressions directly. You only need sympify/parse_expr if you are parsing arbitrary strings that you don't know ahead of time (side question, when should we recommend sympify vs. parse_expr?).

Another question is how we can avoid library code using sympify instead of _sympify. We might want to have some checks, because contributions often just use sympify without realizing that _sympify is better. We should also try to make more classes call sympify automatically in the superclass so it isn't explicitly needed (and for functions, we can use a decorator helper).

**Comment by oscarbenjamin** [comment]
I think if we remove all use of sympify in the codebase then contributors will mostly get the idea.

This is exactly the kind of thing that I would want to have a SymPEP for. It would be good to have a document that explains:

1. Why string input is bad
2. What the deprecation and removal plan is
3. How users can update their code
4. What contributors should use in the sympy codebase and why

The key thing that a SymPEP can communicate unlike docs is *why* a change was made.
